File: apps/web/src/lib/theme.ts
/**
 * Resolve a brand's `BrandColors` into concrete drawing colors.
 *
 * Mirror of the renderer's `Theme.resolve` (packages/renderer/.../theme.py):
 * the same defaults and override→fallback rules, so the contrast panel reflects
 * exactly the colors the PDF is drawn with. Keep the two in sync.
 */

import type { BrandColors } from "./types";

/** Defaults for the optional-but-always-filled slots (see config.py BrandColors). */
const DEFAULTS = {
    text: "#1A1A1A",
    text_soft: "#555555",
    line: "#D6DEE5",
    bg_soft: "#F4F7FA",
    quote_bg: "#EAF4FB",
    white: "#FFFFFF",
} as const;

export interface ResolvedTheme {
    dark: string;
    light: string;
    mid: string;
    text: string;
    text_soft: string;
    line: string;
    bg_soft: string;
    quote_bg: string;
    white: string;
    table_header: string;
    header_rule: string;
    footer_rule: string;
    page_number: string;
    heading: string;
    heading_h3: string;
}

/** Treat empty/whitespace as absent, matching the renderer's `if value` pick(). */
function pick(value: string | undefined, fallback: string): string {
    return value && value.trim() ? value : fallback;
}

/**
 * Expand a (possibly sparse) BrandColors into an explicit set of the nine
 * "base" colors, resolving their defaults. The optional role overrides
 * (table_header_bg, heading, …) are preserved verbatim rather than filled, so
 * a round-trip through the editor never disturbs the heading→h3 fallback split.
 */
export function fullColors(c: BrandColors): BrandColors {
    const t = resolveTheme(c);
    return {
        primary_dark: t.dark,
        primary_light: t.light,
        primary_mid: t.mid,
        text: t.text,
        text_soft: t.text_soft,
        line: t.line,
        bg_soft: t.bg_soft,
        quote_bg: t.quote_bg,
        white: t.white,
        table_header_bg: c.table_header_bg,
        header_rule: c.header_rule,
        footer_rule: c.footer_rule,
        heading: c.heading,
        page_number: c.page_number,
    };
}

export function resolveTheme(c: BrandColors): ResolvedTheme {
    const dark = c.primary_dark;
    const light = c.primary_light;
    const mid = c.primary_mid;
    const line = pick(c.line, DEFAULTS.line);

    return {
        dark,
        light,
        mid,
        text: pick(c.text, DEFAULTS.text),
        text_soft: pick(c.text_soft, DEFAULTS.text_soft),
        line,
        bg_soft: pick(c.bg_soft, DEFAULTS.bg_soft),
        quote_bg: pick(c.quote_bg, DEFAULTS.quote_bg),
        white: pick(c.white, DEFAULTS.white),
        table_header: pick(c.table_header_bg, dark),
        header_rule: pick(c.header_rule, light),
        footer_rule: pick(c.footer_rule, line),
        page_number: pick(c.page_number, light),
        heading: pick(c.heading, dark),
        heading_h3: pick(c.heading, mid),
    };
}

File: packages/renderer/layout_studio_renderer/theme.py
"""Resolves a brand's `BrandColors` into a complete set of concrete drawing
colors.

This is the single place the per-brand override fallback rules live: every
semantic slot below is filled here, so the renderer's draw code reads a ready
color (`self.col_heading`) instead of falling back inline
(`self.col_heading or self.col_dark`) at each call site.
"""

from __future__ import annotations

from dataclasses import dataclass

from reportlab.lib.colors import Color, HexColor

from .config import BrandColors


@dataclass(frozen=True)
class Theme:
    """Fully-resolved palette. Direct slots map 1:1 to BrandColors; the rest
    apply the documented fallbacks."""

    # Direct brand colors.
    dark: Color
    light: Color
    mid: Color
    text: Color
    text_soft: Color
    line: Color
    bg_soft: Color
    quote_bg: Color
    white: Color
    # Role slots (override → fallback).
    table_header: Color  # table_header_bg → dark
    header_rule: Color   # header_rule     → light
    footer_rule: Color   # footer_rule     → line
    page_number: Color   # page_number     → light
    heading: Color       # heading         → dark  (h1 / h2)
    heading_h3: Color    # heading         → mid   (h3)

    @classmethod
    def resolve(cls, cols: BrandColors) -> "Theme":
        dark = HexColor(cols.primary_dark)
        light = HexColor(cols.primary_light)
        mid = HexColor(cols.primary_mid)
        line = HexColor(cols.line)

        def pick(value: str | None, default: Color) -> Color:
            return HexColor(value) if value else default

        return cls(
            dark=dark,
            light=light,
            mid=mid,
            text=HexColor(cols.text),
            text_soft=HexColor(cols.text_soft),
            line=line,
            bg_soft=HexColor(cols.bg_soft),
            quote_bg=HexColor(cols.quote_bg),
            white=HexColor(cols.white),
            table_header=pick(cols.table_header_bg, dark),
            header_rule=pick(cols.header_rule, light),
            footer_rule=pick(cols.footer_rule, line),
            page_number=pick(cols.page_number, light),
            heading=pick(cols.heading, dark),
            heading_h3=pick(cols.heading, mid),
        )
